src/guiapi/src/window.cpp

```diff
@@ -1,351 +1,215 @@
-            } else if (window->cls->draw)
-                window->cls->draw(window);
-        }
+window_t *window_t::GetNextEnabled() const {
+    if (next) {
+        return (next->IsEnabled()) ? next : next->GetNextEnabled();
+    } else {
+        return nullptr;
+    }
 }
 
-void window_t::Invalidate() {
-    f_invalid = 1;
-    gui_invalidate();
+/*
+window_t *window_t::GetPrevEnabled() const {
+    if (prev){
+        return (prev->IsEnabled()) ? prev : prev->GetPrevEnabled();
+    } else {
+        return nullptr;
+    }
 }
+*/
 
-void window_validate_children(int16_t id) {
-    window_t *window;
-    int16_t id_child;
-    for (id_child = 0; id_child < WINDOW_MAX_WINDOWS; id_child++)
-        if (((window = windows[id_child]) != 0) && (window->id_parent == id))
-            window->f_invalid = 0;
+window_t *window_t::GetParent() const {
+    return parent;
 }
 
-void window_invalidate_children(int16_t id) {
-    window_t *window;
-    int16_t id_child;
-    for (id_child = 0; id_child < WINDOW_MAX_WINDOWS; id_child++)
-        if (((window = windows[id_child]) != 0) && (window->id_parent == id))
-            window->f_invalid = 1;
-    gui_invalidate();
+void window_t::Draw() {
+    draw();
 }
 
-void window_t::SetFocus() {
-    if (!f_visible || !f_enabled)
-        return;
-
-    if (window_focused_ptr) {
-        window_focused_ptr->f_focused = 0;
-        window_focused_ptr->f_invalid = 1;
-        if (window_focused_ptr->event)
-            window_focused_ptr->event(window_focused_ptr, WINDOW_EVENT_FOCUS0, 0);
+void window_t::draw() {
+    if (IsInvalid() && IsVisible() && rect.w && rect.h) {
+        unconditionalDraw();
+        Validate();
     }
-    window_focused_ptr = this;
-    f_focused = 1;
-    f_invalid = 1;
-    if (event)
-        event(this, WINDOW_EVENT_FOCUS1, 0);
-    gui_invalidate();
 }
 
-void window_t::SetCapture() {
-
-    if (f_visible && f_enabled && event) {
-        if (window_capture_ptr) {
-            window_capture_ptr->f_capture = 0;
-            if (window_capture_ptr->event)
-                window_capture_ptr->event(window_capture_ptr, WINDOW_EVENT_CAPT_0, 0);
-        }
-        window_capture_ptr = this;
-        f_capture = 1;
-        event(this, WINDOW_EVENT_CAPT_1, 0);
-        gui_invalidate();
-    }
+//window does not support subwindow elements, but window_frame does
+void window_t::RegisterSubWin(window_t *win) {
 }
 
-void window_t::Show() {
-    if (!f_visible) {
-        f_visible = 1;
-        Invalidate();
-    }
+void window_t::unconditionalDraw() {
+    display::FillRect(rect, color_back);
 }
 
-void window_t::Hide() {
-    if (f_visible) {
-        f_visible = 0;
-        Invalidate();
-    }
+void window_t::WindowEvent(window_t *sender, uint8_t ev, void *param) {
+    windowEvent(sender, ev, param);
 }
 
-void window_t::SetBackColor(color_t clr) {
-    color_back = clr;
-    Invalidate();
+void window_t::ScreenEvent(window_t *sender, uint8_t ev, void *param) {
```

Comment:
Same method name with different capital letters? This is dangerous or useless.

Reply:
Public methods have Big capitals
they are usually not virtual and call protected virtual method